docs/fixing-warnings.md

```diff
@@ -0,0 +1,189 @@
+{
+    Console.WriteLine(m.Name);
+}
+```
+
+In this example, `Type.GetType` dynamically requests a type with an unknown name, and then prints
+the names of all of its methods. Because there's no way to know at publish time what type name is
+going to be used, there's no way for the trimmer to know which type to preserve in the output.
+It's very likely that this code could have worked before trimming (as long as the input is
+something known to exist in the target framework), but would probably produce a null reference
+exception after trimming (due to `Type.GetType` returning null).
+
+This is a frustrating situation. Trimming often works just fine but occasionally it can produce
+breaking behavior, sometimes in rare code paths, and it can be very hard to trace down the cause.
+
+For .NET 6 we want to bring a new feature to trimming: trim warnings. Trim warnings happen
+because the trimmer sees a call which may access other code in the app but the trimmer can't
+determine which code. This could mean that the trimmer would trim away code which is used at
+runtime.
+
+## Reacting to trim warnings
+
+Trim warnings are meant to bring predictability to trimming. The problem with trimming is that some
+code patterns can depend on code in a way that isn't understandable by the trimmer. Whenever the
+trimmer encounters code like that, that's when you should expect a trim warning.
+
+There are two big categories of warnings which you will likely see:
+
+ 1. `RequiresUnreferencedCode`
+ 2. `DynamicallyAccessedMembers`
+
+### RequiresUnreferencedCode
+
+`RequiresUnreferencedCode` is simple: it's an attribute that can be placed on methods to indicate
+that the method is not trim-compatible, meaning that it might use reflection or some other
+mechanism to access code that may be trimmed away. This attribute is used when it's not possible
+for the trimmer to understand what's necessary, and a blanket warning is needed. This would often
+be true for methods which use the C# `dynamic` keyword, `Assembly.LoadFrom`, or other runtime
+code generation technologies.
+An example would be:
+
+```C#
+[RequiresUnreferencedCode("Use 'MethodFriendlyToTrimming' instead")]
+void MethodWithAssemblyLoad() { ... }
+
+void TestMethod()
+{
+    // IL2026: Using method 'MethodWithAssemblyLoad' which has 'RequiresUnreferencedCodeAttribute'
+    // can break functionality when trimming application code. Use 'MethodFriendlyToTrimming' instead.
+    MethodWithAssemblyLoad();
+}
+```
+
+There aren't many workarounds for `RequiresUnreferencedCode`. The best way is to avoid calling
+the method at all when trimming and use something else which is trim-compatible. If you're
+writing a library and it's not in your control whether or not to call the method and you just
+want to communicate to *your* caller, you can also add `RequiresUnreferencedCode` to your own
+method. This silences all trimming warnings in your code, but will produce a warning whenever
```

Comment:
Why does this produce a warning rather than an error? If the method is marked that its not going to work, and is called, then why only warn?

Reply:
Because in almost all cases the method might work in some cases and might not in others. For example `XmlSerializer.Serialize(object o)` - it's definitely not trim compatible, but if you make sure that the type passed in is preserved enough - it will work just fine.

We use this attribute basically to say "If you ignore this, you're on your own and will have to test at runtime".

There are also cases where the method is sooo close to being trim compatible except for some weird special case - we can't mark it as compatible (since it's not 100%), but it almost always is.

And as with any other trim analysis - the linker plays it super safe, the annotations do that as well. We don't want to block people completely if we are too careful.

Reply:
Pretty much everything in trimming is a warning, because many things just happen to work. For instance, if you call a method that has RUC does a lot of reflection on its parameter, but the parameter type is in user code (and therefore untrimmed) and nothing important from the framework was removed, it all just works.

Of course, this is just luck, which is why the warning is there.